File: c/otterop/datastructure/linked_list_node.c
#include <otterop/datastructure/int/linked_list_node.h>
#include <gc.h>
#include <stdlib.h>

typedef struct otterop_datastructure_LinkedListNode_s {
    otterop_datastructure_LinkedList_t *list;
    otterop_datastructure_LinkedListNode_t *prev;
    otterop_datastructure_LinkedListNode_t *next;
    void *value;
} otterop_datastructure_LinkedListNode_t;




otterop_datastructure_LinkedListNode_t *otterop_datastructure_LinkedListNode_new(void *value) {
    otterop_datastructure_LinkedListNode_t *self = GC_malloc(sizeof(otterop_datastructure_LinkedListNode_t));
    self->prev = NULL;
    self->next = NULL;
    self->value = value;
    return self;
}

otterop_datastructure_LinkedList_t *otterop_datastructure_LinkedListNode_list(otterop_datastructure_LinkedListNode_t *self) {
    return self->list;
}

void otterop_datastructure_LinkedListNode_set_list(otterop_datastructure_LinkedListNode_t *self, otterop_datastructure_LinkedList_t *list) {
    self->list = list;
}

otterop_datastructure_LinkedListNode_t *otterop_datastructure_LinkedListNode_prev(otterop_datastructure_LinkedListNode_t *self) {
    return self->prev;
}

void otterop_datastructure_LinkedListNode_set_prev(otterop_datastructure_LinkedListNode_t *self, otterop_datastructure_LinkedListNode_t *node) {
    self->prev = node;
}

otterop_datastructure_LinkedListNode_t *otterop_datastructure_LinkedListNode_next(otterop_datastructure_LinkedListNode_t *self) {
    return self->next;
}

void otterop_datastructure_LinkedListNode_set_next(otterop_datastructure_LinkedListNode_t *self, otterop_datastructure_LinkedListNode_t *node) {
    self->next = node;
}

void *otterop_datastructure_LinkedListNode_value(otterop_datastructure_LinkedListNode_t *self) {
    return self->value;
}

File: go/datastructure/linkedlistnode.go
package datastructure



type LinkedListNode[T any] struct {
    list *LinkedList[T]
    prev *LinkedListNode[T]
    next *LinkedListNode[T]
    value T
}




func LinkedListNodeNew[T any](value T) *LinkedListNode[T] {
    this := new(LinkedListNode[T])
    this.prev = nil
    this.next = nil
    this.value = value
    return this
}

func (this *LinkedListNode[T]) List() *LinkedList[T] {
    return this.list
}

func (this *LinkedListNode[T]) setList(list *LinkedList[T])  {
    this.list = list
}

func (this *LinkedListNode[T]) Prev() *LinkedListNode[T] {
    return this.prev
}

func (this *LinkedListNode[T]) setPrev(node *LinkedListNode[T])  {
    this.prev = node
}

func (this *LinkedListNode[T]) Next() *LinkedListNode[T] {
    return this.next
}

func (this *LinkedListNode[T]) setNext(node *LinkedListNode[T])  {
    this.next = node
}

func (this *LinkedListNode[T]) Value() T {
    return this.value
}